### PR TITLE
Lit 2.0: convert run-async directive to class-based syntax

### PR DESCRIPTION
**Pull request by dlockhart**
Building on the new unit tests, this converts the directive to follow the new Lit 2.0 class-based syntax:
https://lit.dev/docs/templates/custom-directives/#async-directives

The big change with the class-based directives is that a `WeakMap` is no longer needed for state, which can instead move into the class. This simplifies things a bit in that you always have access to that state in the callbacks and don't need to keep grabbing it and confirming that it's valid.

Things also get split out into `render()` (which can run on the server) and `update()` which does not. So that's where I put the firing of the `pending-state` event.

### CODE REVIEW

**Comment by dlockhart** [comment]
From Dave B in #1565:
> Besides the unit tests, for the conversion, I would also be tempted to manually check the demo for the async container mixin, and also the dialog demo. Those demos happen to contain usages of runAsync.

Yep for sure, tested these a lot while developing this.

File: directives/run-async/run-async.js [hunk 1]
```diff
@@ -11,16 +16,107 @@
...
+					}
+				}
+			});
+		}
+
+		if ((this.state === 'pending') && typeof pending === 'function') {
+			return pending();
+		}
+
+		return noChange;
```
Comment (dbatiste) [inline directives/run-async/run-async.js:98]: Interesting... https://lit.dev/docs/api/custom-directives/#noChange. 😄

**Review by dbatiste — APPROVED** [review]
Looks good!  So for the most part, it looks like it's following the original logic fairly closely with the key differences as you highlighted.

I wonder if it still needs to wait a microtask for the initial render to complete before dispatching the event.  If that's not needed, it looks like the `async() => { ... })();` wrapper would no longer be needed.

**Comment by dlockhart** [comment]
> I wonder if it still needs to wait a microtask for the initial render to complete before dispatching the event. If that's not needed, it looks like the async() => { ... })(); wrapper would no longer be needed.

I just played around with this, and it looks like it can be removed. Although now the first time it runs happens before there's a host so `part.parentNode.host` isn't set yet. I added some checking for that and things seem to be working.

File: directives/run-async/run-async.js [hunk 2]
```diff
@@ -11,16 +16,103 @@
...
+
+	}
+
+	update(part, [key, task, templates, options]) {
+		const renderResult = this.render(key, task, templates);
+		const directiveOptions = Object.assign({ pendingState: true }, options);
+		if (directiveOptions.pendingState && this.key === key && this.state === 'pending') {
+			const element = part.parentNode.nodeType === Node.ELEMENT_NODE ? part.parentNode : part.parentNode.host;
+			if (element) {
```
Comment (dbatiste) [inline directives/run-async/run-async.js:107]: Could `element` ever be `null` or `undefined` and `update` not run again, such that the event never gets dispatched?

Comment (dbatiste) [reply directives/run-async/run-async.js:107]: Or, what if the async task immediately returns a resolve promise and the state goes from pending -> success?  Could that happen before this `update` method runs, where dispatching the event relies on the state change?  I'm not sure what the behaviour of that was before.

Comment (dlockhart) [reply directives/run-async/run-async.js:107]: I think that's what's happening in Firefox & legacy-Edge in the failing tests. :(

**Review by dbatiste — APPROVED** [review]

**Pull request merged**